I have a error  with  CHAN: REORGANIZE:

sunbodeMBP:bin sunbo$ ./btcd
2018-06-14 14:14:29.253 [INF] BTCD: Version 0.12.0-beta
2018-06-14 14:14:29.253 [INF] BTCD: Loading block database from '/Users/sunbo/Library/Application Support/Btcd/data/mainnet/blocks_ffldb'
2018-06-14 14:14:29.326 [INF] BTCD: Block database loaded
2018-06-14 14:14:29.338 [INF] INDX: cf index is enabled
2018-06-14 14:14:29.338 [INF] CHAN: Loading block index...
2018-06-14 14:14:31.483 [INF] CHAN: Chain state (height 384560, hash 000000000000000005a15c23c9f1d67a8e5d93944e73fb74a4177c0611992d02, totaltx 93547834, work 14942772535223485067948946)
2018-06-14 14:14:31.497 [INF] RPCS: RPC server listening on 127.0.0.1:8334
2018-06-14 14:14:31.497 [INF] RPCS: RPC server listening on [::1]:8334
2018-06-14 14:14:31.504 [INF] AMGR: Loaded 278 addresses from file '/Users/sunbo/Library/Application Support/Btcd/data/mainnet/peers.json'
2018-06-14 14:14:31.506 [INF] CMGR: 40 addresses found from DNS seed seed.bitcoin.sipa.be
2018-06-14 14:14:31.513 [INF] CMGR: Server listening on [::]:8333
2018-06-14 14:14:31.515 [INF] CMGR: 38 addresses found from DNS seed dnsseed.bitcoin.dashjr.org
2018-06-14 14:14:31.515 [INF] CMGR: 38 addresses found from DNS seed seed.bitcoin.jonasschnelli.ch
2018-06-14 14:14:31.517 [INF] CMGR: Server listening on 0.0.0.0:8333
2018-06-14 14:14:31.580 [INF] CMGR: 33 addresses found from DNS seed dnsseed.bluematt.me
2018-06-14 14:14:31.635 [INF] CMGR: 41 addresses found from DNS seed seed.bitcoinstats.com
2018-06-14 14:14:31.676 [INF] CMGR: 40 addresses found from DNS seed seed.bitnodes.io
2018-06-14 14:14:32.527 [INF] SYNC: New valid peer 64.99.224.88:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:32.527 [INF] SYNC: Syncing to block height 527381 from peer 64.99.224.88:8333
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 180.181.118.167:8333 (outbound) (/Satoshi:0.16.0(MY_BALLS)/)
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 73.250.151.156:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 82.95.192.186:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 103.82.56.24:8333 (outbound) (/Satoshi:0.15.99/)
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 155.143.140.186:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:32.528 [INF] SYNC: New valid peer 155.143.140.186:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:32.530 [INF] SYNC: Lost peer 82.95.192.186:8333 (outbound)
2018-06-14 14:14:33.221 [INF] SYNC: New valid peer 75.24.121.227:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:33.244 [INF] SYNC: New valid peer 176.9.158.150:8333 (outbound) (/Satoshi:0.16.0/)
2018-06-14 14:14:35.688 [INF] CHAN: REORGANIZE: Block 0000000000000000063986c60f4b617bf7310b58cdde652ceddd3addc62eb2cc is causing a reorganize.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x135b6d7]

goroutine 41 [running]:
github.com/btcsuite/btcd/blockchain.(*BlockChain).reorganizeChain(0xc420073e00, 0xc428b3a2d0, 0xc428b3a2a0, 0xc4265000d0, 0x1)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/blockchain/chain.go:1009 +0x12c7
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBestChain(0xc420073e00, 0xc428a0f8f0, 0xc4289fccb0, 0x0, 0x0, 0x0, 0xc429455ad8)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/blockchain/chain.go:1129 +0xa5f
github.com/btcsuite/btcd/blockchain.(*BlockChain).maybeAcceptBlock(0xc420073e00, 0xc4289fccb0, 0xffffffff00000000, 0x0, 0x0, 0x1ac8100)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/blockchain/accept.go:79 +0x207
github.com/btcsuite/btcd/blockchain.(*BlockChain).ProcessBlock(0xc420073e00, 0xc4289fccb0, 0xc400000000, 0xc426980000, 0x0, 0x0)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/blockchain/process.go:228 +0x2b5
github.com/btcsuite/btcd/netsync.(*SyncManager).handleBlockMsg(0xc420162420, 0xc4262c0fa0)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/netsync/manager.go:569 +0x205
github.com/btcsuite/btcd/netsync.(*SyncManager).blockHandler(0xc420162420)
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/netsync/manager.go:1166 +0x548
created by github.com/btcsuite/btcd/netsync.(*SyncManager).Start
	/Users/sunbo/Desktop/PubGo/src/github.com/btcsuite/btcd/netsync/manager.go:1388 +0xe3

Mine is dead as well:

```
2018-06-21 13:17:16.761 [INF] CHAN: REORGANIZE: Block 0000000000000000010dd18650065dba688100ab1c8807e679ed028adcdc9d5c is causing a reorganize.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x75a9d7]

goroutine 34 [running]:
github.com/btcsuite/btcd/blockchain.(*BlockChain).reorganizeChain(0xc4204c7680, 0xc4251104e0, 0xc4251104b0, 0xc42013c620, 0x1)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/blockchain/chain.go:1009 +0x12c7
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBestChain(0xc4204c7680, 0xc421fce0e0, 0xc4220be700, 0x0, 0x0, 0x0, 0xc42ade5ad8)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/blockchain/chain.go:1129 +0xa5f
github.com/btcsuite/btcd/blockchain.(*BlockChain).maybeAcceptBlock(0xc4204c7680, 0xc4220be700, 0xffffffff00000000, 0x0, 0x0, 0xec77a0)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/blockchain/accept.go:79 +0x207
github.com/btcsuite/btcd/blockchain.(*BlockChain).ProcessBlock(0xc4204c7680, 0xc4220be700, 0xc400000000, 0xc429e70000, 0x0, 0x0)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/blockchain/process.go:228 +0x2b5
github.com/btcsuite/btcd/netsync.(*SyncManager).handleBlockMsg(0xc4201642c0, 0xc426667e00)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/netsync/manager.go:569 +0x205
github.com/btcsuite/btcd/netsync.(*SyncManager).blockHandler(0xc4201642c0)
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/netsync/manager.go:1166 +0x548
created by github.com/btcsuite/btcd/netsync.(*SyncManager).Start
        /home/ubuntu/gocode/src/github.com/btcsuite/btcd/netsync/manager.go:1388 +0xe3
```

--- a/blockchain/chain.go
+++ b/blockchain/chain.go
@@ -1,4 +1,5 @@
-// Copyright (c) 2013-2017 The btcsuite developers
+// Copyright (c) 2013-2018 The btcsuite developers
+// Copyright (c) 2015-2018 The Decred developers
 // Use of this source code is governed by an ISC
 // license that can be found in the LICENSE file.
 
@@ -819,6 +820,38 @@ func countSpentOutputs(block *btcutil.Block) int {
 //
 // This function MUST be called with the chain state lock held (for writes).
 func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error {
+	// Nothing to do if no reorganize nodes were provided.
+	if detachNodes.Len() == 0 && attachNodes.Len() == 0 {
+		return nil
+	}
+
+	// Ensure the provided nodes match the current best chain.
+	tip := b.bestChain.Tip()
+	if detachNodes.Len() != 0 {
+		firstDetachNode := detachNodes.Front().Value.(*blockNode)
+		if firstDetachNode.hash != tip.hash {
+			return AssertError(fmt.Sprintf("reorganize nodes to detach are "+
+				"not for the current best chain -- first detach node %v, "+
+				"current chain %v", &firstDetachNode.hash, &tip.hash))
+		}
+	}
+
+	// Ensure the provided nodes are for the same fork point.
+	if attachNodes.Len() != 0 && detachNodes.Len() != 0 {
+		firstAttachNode := attachNodes.Front().Value.(*blockNode)
+		lastDetachNode := detachNodes.Back().Value.(*blockNode)
+		if firstAttachNode.parent.hash != lastDetachNode.parent.hash {
+			return AssertError(fmt.Sprintf("reorganize nodes do not have the "+
+				"same fork point -- first attach parent %v, last detach "+
+				"parent %v", &firstAttachNode.parent.hash,
+				&lastDetachNode.parent.hash))
+		}
+	}
+
+	// Track the old and new best chains heads.
+	oldBest := tip
+	newBest := tip
+
 	// All of the blocks to detach and related spend journal entries needed
 	// to unspend transaction outputs in the blocks being disconnected must
 	// be loaded from the database during the reorg check phase below and
@@ -833,7 +866,7 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 	// database and using that information to unspend all of the spent txos
 	// and remove the utxos created by the blocks.
 	view := NewUtxoViewpoint()
-	view.SetBestHash(&b.bestChain.Tip().hash)
+	view.SetBestHash(&oldBest.hash)
 	for e := detachNodes.Front(); e != nil; e = e.Next() {
 		n := e.Value.(*blockNode)
 		var block *btcutil.Block
@@ -844,6 +877,11 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 		})
 		if err != nil {
 			return err
+		}
+		if n.hash != *block.Hash() {
+			return AssertError(fmt.Sprintf("detach block node hash %v (height "+
+				"%v) does not match previous parent block hash %v", &n.hash,
+				n.height, block.Hash()))
 		}
 
 		// Load all of the utxos referenced by the block that aren't
@@ -872,6 +910,15 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 		if err != nil {
 			return err
 		}
+
+		newBest = n.parent
+	}
+
+	// Set the fork point only if there are nodes to attach since otherwise
+	// blocks are only being disconnected and thus there is no fork point.
+	var forkNode *blockNode
+	if attachNodes.Len() > 0 {
+		forkNode = newBest
 	}
 
 	// Perform several checks to verify each block that needs to be attached
@@ -886,16 +933,8 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 	// at least a couple of ways accomplish that rollback, but both involve
 	// tweaking the chain and/or database.  This approach catches these
 	// issues before ever modifying the chain.
-	var validationError error
 	for e := attachNodes.Front(); e != nil; e = e.Next() {
 		n := e.Value.(*blockNode)
-
-		// If any previous nodes in attachNodes failed validation,
-		// mark this one as having an invalid ancestor.
-		if validationError != nil {
-			b.index.SetStatusFlags(n, statusInvalidAncestor)
-			continue
-		}
 
 		var block *btcutil.Block
 		err := b.db.View(func(dbTx database.Tx) error {
@@ -922,6 +961,8 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 			if err != nil {
 				return err
 			}
+
+			newBest = n
 			continue
 		}
 
@@ -929,23 +970,24 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 		// thus will not be generated.  This is done because the state
 		// is not being immediately written to the database, so it is
 		// not needed.
+		//
+		// In the case the block is determined to be invalid due to a
+		// rule violation, mark it as invalid and mark all of its
+		// descendants as having an invalid ancestor.
 		err = b.checkConnectBlock(n, block, view, nil)
 		if err != nil {
-			// If the block failed validation mark it as invalid, then
-			// continue to loop through remaining nodes, marking them as
-			// having an invalid ancestor.
 			if _, ok := err.(RuleError); ok {
 				b.index.SetStatusFlags(n, statusValidateFailed)
-				validationError = err
-				continue
+				for de := e.Next(); de != nil; de = de.Next() {
+					dn := de.Value.(*blockNode)
+					b.index.SetStatusFlags(dn, statusInvalidAncestor)
+				}
 			}
 			return err
 		}
 		b.index.SetStatusFlags(n, statusValid)
-	}
 
-	if validationError != nil {
-		return validationError
+		newBest = n
 	}
 
 	// Reset the view for the actual connection code below.  This is
@@ -1014,12 +1056,14 @@ func (b *BlockChain) reorganizeChain(detachNodes, attachNodes *list.List) error 
 
 	// Log the point where the chain forked and old and new best chain
 	// heads.
-	firstAttachNode := attachNodes.Front().Value.(*blockNode)
-	firstDetachNode := detachNodes.Front().Value.(*blockNode)
-	lastAttachNode := attachNodes.Back().Value.(*blockNode)
-	log.Infof("REORGANIZE: Chain forks at %v", firstAttachNode.parent.hash)
-	log.Infof("REORGANIZE: Old best chain head was %v", firstDetachNode.hash)
-	log.Infof("REORGANIZE: New best chain head is %v", lastAttachNode.hash)
+	if forkNode != nil {
+		log.Infof("REORGANIZE: Chain forks at %v (height %v)", forkNode.hash,
+			forkNode.height)
+	}
+	log.Infof("REORGANIZE: Old best chain head was %v (height %v)",
+		&oldBest.hash, oldBest.height)
+	log.Infof("REORGANIZE: New best chain head is %v (height %v)",
+		newBest.hash, newBest.height)
 
 	return nil
 }

